src/utils/transition-manager.js

```diff
@@ -0,0 +1,209 @@
+import assert from 'assert';
+import {
+  viewportLinearInterpolator,
+  extractViewportFrom,
+  VIEWPORT_PROPS
+} from './viewport-transition-utils';
+import MapState from './map-state';
+
+const VIEWPORT_TRANSITION_FREQUENCY = 60; // 60 times per second.
+const VIEWPORT_TRANSITION_EASING_FUNC = t => t;
+
+export const TRANSITION_EVENTS = {
+  BREAK: 1,
+  SNAP_TO_END: 2,
+  IGNORE: 3
+};
+
+const DEFAULT_PROPS = {
+  transitionDuration: 0,
+  transitionInterpolator: viewportLinearInterpolator,
+  transitionEasing: VIEWPORT_TRANSITION_EASING_FUNC,
+  transitionInterruption: TRANSITION_EVENTS.BREAK,
+  onTransitionStart: () => {},
+  onTransitionInterrupt: () => {},
+  onTransitionEnd: () => {}
+};
+
+export default class TransitionManager {
+  constructor(props) {
+    this.props = props;
+
+    this.transitionContext = {
+      time: 0,
+      delta: 0,
+      interval: null,
+      viewport: null,
+      startViewport: null,
+      endViewport: null
+    };
+  }
+
+  // Returns current transitioned viewport.
+  getTransionedViewport() {
+    return this.transitionContext.viewport;
+  }
+
+  // Process the vewiport change, either ignore or trigger a new transiton.
+  processViewportChange(nextProps) {
+    const isTransitionInProgress = this._isTransitionInProgress();
+    const isTransitionEnabled = this._isTransitionEnabled(nextProps);
+
+    // NOTE: Be cautious re-ordering statements in this function.
+    if (!isTransitionEnabled && !isTransitionInProgress ||
+      this._shouldIgnoreViewportChange(nextProps)) {
+      this.props = nextProps;
+      return;
+    }
+
+    if (isTransitionEnabled) {
+      const currentViewport = this.transitionContext.viewport || extractViewportFrom(this.props);
+      const endViewport = this.transitionContext.endViewport;
+
+      const startViewport = this._shouldSnapToEnd() ?
+        (endViewport || currentViewport) :
+        currentViewport;
+
+      this._triggerTransition(startViewport, nextProps);
+
+      if (isTransitionInProgress) {
+        this.props.onTransitionInterrupt();
+      }
+      nextProps.onTransitionStart();
+    } else if (isTransitionInProgress) {
+      this.props.onTransitionInterrupt();
+      this._endTransition();
+    }
+
+    this.props = nextProps;
+  }
+
+  // Helper methods
+
+  /* eslint-disable max-depth */
+  _areViewportsEqual(startViewport, endViewport) {
+    for (const p of VIEWPORT_PROPS) {
+      if (Array.isArray(startViewport[p])) {
+        for (let i = 0; i < startViewport[p].length; ++i) {
+          if (startViewport[p][i] !== endViewport[p][i]) {
+            return false;
+          }
+        }
+      } else if (startViewport[p] !== endViewport[p]) {
+        return false;
+      }
+    }
+    return true;
+  }
+  /* eslint-enable max-depth */
+
+  _createTransitionInterval(delay) {
+    if (this.transitionContext.interval) {
+      clearInterval(this.transitionContext.interval);
+    }
+    return setInterval(() => this._updateViewport(), delay);
```

Comment:
This is copied from deck.gl - should be changed to using `requestAnimationFrame`.